src/Microsoft.Data/DataFrame.cs

```diff
@@ -260,6 +261,77 @@ public DataFrame Join(DataFrame other, string leftSuffix = "_left", string right
             return ret;
         }
 
+        public GroupBy GroupBy(string columnName)
+        {
+            int columnIndex = _table.GetColumnIndex(columnName);
+            if (columnIndex == -1)
+                throw new ArgumentException($"{columnName} does not exist");
+
+            BaseColumn column = _table.Column(columnIndex);
+
+            switch (column)
```

Comment:
How about instead of a massive switch like this (and not really being able to support new, extension column types), why don't we have a public abstract method on `BaseColumn` that gets called here?

Reply:
>  I was just trying to avoid boxing 

What would be boxed? `BaseColumn` and `GropuBy` are classes and thus reference types. Calling a an abstract method on it won't box anything.

>  the default case is doing that right?

No, the default case is 

```C#
default:
                    Dictionary<object, ICollection<long>> dictionary = column.HashColumnValues<object>();
                    return new GroupBy<object>(this, columnIndex, dictionary);
```

Which means the dictionary is always a `Dictionary<object, ...>`.  If you made a method on BaseColumn that looked like 

```C#
public abstract GroupBy GroupBy(int columnIndex);
```

Then each column can decide how to create the `GroupBy` object and its underlying Dictionary.

Reply:
> What would be boxed? BaseColumn and GropuBy are classes and thus reference types. Calling a an abstract method on it won't box anything.

Right right, but that's not what I meant. So, the way it is set up at the moment, let's say I have a `PrimitiveColumn<int> intColumn`. The `GroupBy` method will call `intColumn.HashColumnValues<int>` to produce a `Dictionary<int...`. That's ok, no boxing involved. If instead, I had only the `default` case of `column.HashColumnValues<object>`(where `column` is really `intColumn` and 'object' is really `int`), that would involve boxing when going in and out of the dictionary right? Anyway, I set it up this way because I didn't want the `BaseColumn` objects to have a reference to their parent `DataFrame` object. With your API suggestion, maybe I can still do that. Let me see